File: kreta_client.py
"""In-memory, read-only KRÉTA API client for the local MCP server."""

from __future__ import annotations

import threading
import time
from typing import Any

import requests

from kreta_smoke_test import (
    CLIENT_ID,
    MOBILE_API_KEY,
    MOBILE_USER_AGENT,
    TIMEOUT_SECONDS,
    TOKEN_URL,
    Credentials,
    SmokeTestError,
    load_credentials,
    login,
    revoke_refresh_token,
)


class KretaClient:
    """Authenticate with OAuth 2.0 + PKCE and call fixed read-only API routes.

    Access and refresh tokens live only in memory. The class refreshes an
    expiring access token and revokes the refresh token when it is closed.
    """

    def __init__(
        self,
        credentials: Credentials | None = None,
        session: requests.Session | None = None,
    ) -> None:
        self.credentials = credentials or load_credentials()
        self.session = session or requests.Session()
        self._owns_session = session is None
        self._token: dict[str, Any] | None = None
        self._expires_at = 0.0
        self._lock = threading.RLock()
        self.base_url = (
            f"https://{self.credentials.institute_code.lower()}.e-kreta.hu/"
            "ellenorzo/v3/"
        )

    def authenticate(self, *, force_refresh: bool = False) -> None:
        """Ensure a usable access token is available."""
        with self._lock:
            if (
                not force_refresh
                and self._token
                and self._token.get("access_token")
                and time.monotonic() < self._expires_at - 60
            ):
                return

            refresh_token = str((self._token or {}).get("refresh_token", ""))
            if refresh_token:
                try:
                    self._token = self._refresh(refresh_token)
                    self._set_expiry(self._token)
                    return
                except SmokeTestError:
                    # A rejected/expired refresh token is safe to replace with a
                    # fresh Authorization Code + PKCE login.
                    self._token = None
                    self._expires_at = 0.0

            self._token = login(self.session, self.credentials)
            self._set_expiry(self._token)

    def _refresh(self, refresh_token: str) -> dict[str, Any]:
        try:
            response = self.session.post(
                TOKEN_URL,
                data={
                    "grant_type": "refresh_token",
                    "refresh_token": refresh_token,
                    "client_id": CLIENT_ID,
                },
                timeout=TIMEOUT_SECONDS,
            )
            response.raise_for_status()
            token = response.json()
        except (requests.RequestException, ValueError) as exc:
            raise SmokeTestError("A KRÉTA hozzáférési token frissítése sikertelen.") from exc

        if not token.get("access_token"):
            raise SmokeTestError("A frissített KRÉTA-tokenből hiányzik az access token.")
        if not token.get("refresh_token"):
            token["refresh_token"] = refresh_token
        return token

    def _set_expiry(self, token: dict[str, Any]) -> None:
        try:
            expires_in = max(1, int(token.get("expires_in", 300)))
        except (TypeError, ValueError):
            expires_in = 300
        self._expires_at = time.monotonic() + expires_in

    def get_json(self, path: str, params: Any = None) -> Any:
        """GET a relative Student API path and return its decoded JSON."""
        normalized_path = path.lstrip("/")
        if (
            not normalized_path
            or "://" in normalized_path
            or ".." in normalized_path.split("/")
        ):
            raise ValueError("Csak relatív, rögzített KRÉTA API-útvonal használható.")

        for attempt in range(2):
            self.authenticate(force_refresh=attempt == 1)
            access_token = str((self._token or {}).get("access_token", ""))
            try:
                response = self.session.get(
                    self.base_url + normalized_path,
                    headers={
                        "Authorization": f"Bearer {access_token}",
                        "User-Agent": MOBILE_USER_AGENT,
                        "apiKey": MOBILE_API_KEY,
                        "Accept": "application/json",
                    },
                    params=params,
                    timeout=TIMEOUT_SECONDS,
                )
            except requests.RequestException as exc:
                raise SmokeTestError("A KRÉTA API-kérés hálózati hibával leállt.") from exc

            if response.status_code == 401 and attempt == 0:
                continue
            if not 200 <= response.status_code < 300:
                raise SmokeTestError(
                    f"A KRÉTA API HTTP {response.status_code} választ adott."
                )
            if response.status_code == 204 or not response.content:
                return None
            try:
                return response.json()
            except ValueError as exc:
                raise SmokeTestError("A KRÉTA API nem JSON választ adott.") from exc

        raise SmokeTestError("A KRÉTA munkamenet nem frissíthető.")

    def status(self) -> dict[str, Any]:
        """Authenticate and return non-sensitive connection metadata."""
        self.authenticate()
        return {
            "authenticated": True,
            "label": self.credentials.label or None,
            "institution": self.credentials.institute_code,
            "authentication": "OAuth 2.0 Authorization Code + PKCE",
            "read_only": True,
            "token_storage": "memory_only",
        }

    def close(self) -> None:
        """Revoke the in-memory refresh token and close the HTTP session."""
        with self._lock:
            refresh_token = str((self._token or {}).get("refresh_token", ""))
            self._token = None
            self._expires_at = 0.0
            if refresh_token:
                revoke_refresh_token(self.session, refresh_token)
            if self._owns_session:
                self.session.close()

    def __enter__(self) -> KretaClient:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

File: desktop/server/kreta-client.js
/**
 * Memóriában élő, csak olvasó KRÉTA API-kliens (a kreta_client.py portja).
 *
 * A tokenek csak memóriában élnek; a lejáró access tokent a kliens frissíti,
 * záráskor a refresh tokent visszavonja. Az egyidejű tool-hívások közös
 * login-t kapnak: az authenticate() egy folyamatban lévő ígéretet oszt meg.
 */

import {
  CookieJar,
  KretaError,
  MOBILE_API_KEY,
  MOBILE_USER_AGENT,
  TIMEOUT_MS,
  login,
  refresh,
  revokeRefreshToken,
} from "./kreta-auth.js";

export class KretaClient {
  constructor(credentials) {
    this.credentials = credentials;
    this.jar = new CookieJar();
    this.token = null;
    this.expiresAt = 0;
    this.pending = null;
    this.baseUrl = `https://${credentials.instituteCode.toLowerCase()}.e-kreta.hu/ellenorzo/v3/`;
  }

  async authenticate({ forceRefresh = false } = {}) {
    if (
      !forceRefresh &&
      this.token?.access_token &&
      Date.now() < this.expiresAt - 60_000
    ) {
      return;
    }
    if (!this.pending) {
      this.pending = this.#renewToken().finally(() => {
        this.pending = null;
      });
    }
    await this.pending;
  }

  async #renewToken() {
    const refreshToken = this.token?.refresh_token ?? "";
    if (refreshToken) {
      try {
        this.#setToken(await refresh(this.jar, refreshToken));
        return;
      } catch (error) {
        if (!(error instanceof KretaError)) throw error;
        // Elutasított/lejárt refresh token: mehet a friss PKCE-bejelentkezés.
        this.token = null;
        this.expiresAt = 0;
      }
    }
    this.#setToken(await login(this.jar, this.credentials));
  }

  #setToken(token) {
    const expiresIn = Math.max(1, Number(token.expires_in) || 300);
    this.token = token;
    this.expiresAt = Date.now() + expiresIn * 1000;
  }

  /**
   * Relatív Student API-útvonal lekérése; a params lehet objektum vagy
   * [kulcs, érték] párok listája (utóbbi ismételt paraméterhez kell).
   */
  async getJson(path, params = null) {
    const normalized = path.replace(/^\/+/, "");
    if (
      !normalized ||
      normalized.includes("://") ||
      normalized.split("/").includes("..")
    ) {
      throw new KretaError("Csak relatív, rögzített KRÉTA API-útvonal használható.");
    }
    const query = new URLSearchParams(params ?? undefined).toString();
    const url = this.baseUrl + normalized + (query ? `?${query}` : "");

    for (let attempt = 0; attempt < 2; attempt += 1) {
      await this.authenticate({ forceRefresh: attempt === 1 });
      let response;
      try {
        response = await fetch(url, {
          headers: {
            Authorization: `Bearer ${this.token?.access_token ?? ""}`,
            "User-Agent": MOBILE_USER_AGENT,
            apiKey: MOBILE_API_KEY,
            Accept: "application/json",
          },
          signal: AbortSignal.timeout(TIMEOUT_MS),
        });
      } catch (cause) {
        throw new KretaError("A KRÉTA API-kérés hálózati hibával leállt.", {
          cause,
        });
      }
      if (response.status === 401 && attempt === 0) continue;
      if (!response.ok) {
        throw new KretaError(`A KRÉTA API HTTP ${response.status} választ adott.`);
      }
      const body = await response.text();
      if (response.status === 204 || !body) return null;
      try {
        return JSON.parse(body);
      } catch {
        throw new KretaError("A KRÉTA API nem JSON választ adott.");
      }
    }
    throw new KretaError("A KRÉTA munkamenet nem frissíthető.");
  }

  /** Bejelentkezik, és nem érzékeny kapcsolati metaadatokat ad vissza. */
  async status() {
    await this.authenticate();
    return {
      authenticated: true,
      label: this.credentials.label || null,
      institution: this.credentials.instituteCode,
      authentication: "OAuth 2.0 Authorization Code + PKCE",
      read_only: true,
      token_storage: "memory_only",
    };
  }

  /** A memóriában lévő refresh token visszavonása. */
  async close() {
    const refreshToken = this.token?.refresh_token ?? "";
    this.token = null;
    this.expiresAt = 0;
    if (refreshToken) await revokeRefreshToken(this.jar, refreshToken);
  }
}
